spanner/transaction.go

```diff
@@ -785,22 +785,36 @@ func (t *ReadOnlyTransaction) begin(ctx context.Context) error {
 		}
 	}()
 	// Retry the BeginTransaction call if a 'Session not found' is returned.
+	nRPCs := uint64(0)
 	for {
+		nRPCs++
 		sh, err = t.sp.takeMultiplexed(ctx)
 		if err != nil {
 			return err
 		}
+
 		t.setSessionEligibilityForLongRunning(sh)
 		sh.updateLastUseTime()
 		var md metadata.MD
-		res, err = sh.getClient().BeginTransaction(contextWithOutgoingMetadata(ctx, sh.getMetadata(), t.disableRouteToLeader), &sppb.BeginTransactionRequest{
+
+		client := sh.getClient()
+		// Firstly set the number of retries as the RPCID.
+		gcl, ok := client.(*grpcSpannerClient)
+		if ok {
+			gcl.setRPCID(nRPCs)
```

Comment:
This seems to confuse two different types of retries:
1. Single RPCs are retried if they receive a retryable error (e.g. `UNAVAILABLE`). This is handled by the gRPC libraries and transparent to the Spanner client. This attempt number should be based on the number of retries that the RPC is being retried by the gRPC libraries.
2. Read/write transactions are retried if Spanner returns an `ABORTED` error. That will cause the entire transaction to be retried. That should not affect the attempt number that is used for an RPC.

Reply:
Yeah I went with an entirely and more direct approach.